RxScala: buffer(timespan,timeshift) raises IllegalArgumentException for timeshift < 1 second

Originally filed by @lJoublanc  in https://github.com/ReactiveX/RxJava/issues/1349:

---

I'm on 0.16. The following code will throw an IllegalArgumentException:

```
scala> val obs = Observable.interval(100 milliseconds).buffer(3 seconds, 500 milliseconds)
obs: rx.lang.scala.Observable[Seq[Long]] = rx.lang.scala.JavaConversions$$anon$1@6f6b1731

scala> obs.subscribe{x =>println(x)}
rx.exceptions.OnErrorNotImplementedException
    at rx.Observable$25.onError(Observable.java:6805)
    at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:124)
    at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:93)
    at rx.Observable$2.call(Observable.java:280)
    at rx.Observable$2.call(Observable.java:268)
    at rx.Observable.subscribe(Observable.java:7076)
    at rx.Observable.protectivelyWrapAndSubscribe(Observable.java:5396)
    at rx.Observable.subscribe(Observable.java:6796)
    at rx.lang.scala.Observable$class.subscribe(Observable.scala:132)
    at rx.lang.scala.JavaConversions$$anon$1.subscribe(JavaConversions.scala:53)
    at .<init>(<console>:14)
    at .<clinit>(<console>)
    at .<init>(<console>:7)
    at .<clinit>(<console>)
    at $print(<console>)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:734)
    at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:983)
    at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:573)
    at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:604)
    at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:568)
    at scala.tools.nsc.interpreter.ILoop.reallyInterpret$1(ILoop.scala:756)
    at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:801)
    at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:713)
    at scala.tools.nsc.interpreter.ILoop.processLine$1(ILoop.scala:577)
    at scala.tools.nsc.interpreter.ILoop.innerLoop$1(ILoop.scala:584)
    at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:587)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply$mcZ$sp(ILoop.scala:878)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:833)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:833)
    at scala.tools.nsc.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:135)
    at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:833)
    at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:83)
    at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:96)
    at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:105)
    at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
Caused by: java.lang.IllegalArgumentException
    at java.util.concurrent.ScheduledThreadPoolExecutor.scheduleAtFixedRate(ScheduledThreadPoolExecutor.java:562)
    at rx.schedulers.ExecutorScheduler.schedulePeriodically(ExecutorScheduler.java:75)
    at rx.operators.ChunkedOperation$TimeBasedChunkCreator.<init>(ChunkedOperation.java:579)
    at rx.operators.OperationBuffer$7.onSubscribe(OperationBuffer.java:360)
    at rx.Observable$1.call(Observable.java:235)
    at rx.Observable$1.call(Observable.java:231)
    at rx.Observable$2.call(Observable.java:272)
    ... 35 more
```

Looking at the implementation of the scheduler, the error is raised when the argument is less than zero. So there appears to be a problem causing the `timeshift` to be rounded down. If you change the `timeshift` to 1 second or larger, it works fine.

--- a/src/main/scala/rx/lang/scala/Observable.scala
+++ b/src/main/scala/rx/lang/scala/Observable.scala
@@ -568,9 +568,9 @@ trait Observable[+T]
    *         a fixed timespan has elapsed.
    */
   def slidingBuffer(timespan: Duration, timeshift: Duration): Observable[Seq[T]] = {
-    val span: Long = timespan.length
-    val shift: Long = timespan.unit.convert(timeshift.length, timeshift.unit)
-    val unit: TimeUnit = timespan.unit
+    val span: Long = timespan.toNanos
+    val shift: Long = timespan.toNanos
+    val unit: TimeUnit = duration.NANOSECONDS
     val oJava: rx.Observable[_ <: java.util.List[_]] = asJavaObservable.buffer(span, shift, unit)
     Observable.jObsOfListToScObsOfSeq(oJava.asInstanceOf[rx.Observable[_ <: java.util.List[T]]])
   }
@@ -592,9 +592,9 @@ trait Observable[+T]
    *         a fixed timespan has elapsed.
    */
   def slidingBuffer(timespan: Duration, timeshift: Duration, scheduler: Scheduler): Observable[Seq[T]] = {
-    val span: Long = timespan.length
-    val shift: Long = timespan.unit.convert(timeshift.length, timeshift.unit)
-    val unit: TimeUnit = timespan.unit
+    val span: Long = timespan.toNanos
+    val shift: Long = timespan.toNanos
+    val unit: TimeUnit = duration.NANOSECONDS
     val oJava: rx.Observable[_ <: java.util.List[_]] = asJavaObservable.buffer(span, shift, unit, scheduler)
     Observable.jObsOfListToScObsOfSeq(oJava.asInstanceOf[rx.Observable[_ <: java.util.List[T]]])
   }
@@ -807,9 +807,9 @@ trait Observable[+T]
    *         a fixed timespan has elapsed.
    */
   def sliding(timespan: Duration, timeshift: Duration): Observable[Observable[T]] = {
-    val span: Long = timespan.length
-    val shift: Long = timespan.unit.convert(timeshift.length, timeshift.unit)
-    val unit: TimeUnit = timespan.unit
+    val span: Long = timespan.toNanos
+    val shift: Long = timespan.toNanos
+    val unit: TimeUnit = duration.NANOSECONDS
     Observable.jObsOfJObsToScObsOfScObs(asJavaObservable.window(span, shift, unit))
       : Observable[Observable[T]] // SI-7818
   }
@@ -831,9 +831,9 @@ trait Observable[+T]
    *         a fixed timespan has elapsed.
    */
   def sliding(timespan: Duration, timeshift: Duration, scheduler: Scheduler): Observable[Observable[T]] = {
-    val span: Long = timespan.length
-    val shift: Long = timespan.unit.convert(timeshift.length, timeshift.unit)
-    val unit: TimeUnit = timespan.unit
+    val span: Long = timespan.toNanos
+    val shift: Long = timespan.toNanos
+    val unit: TimeUnit = duration.NANOSECONDS
     Observable.jObsOfJObsToScObsOfScObs(asJavaObservable.window(span, shift, unit, scheduler))
       : Observable[Observable[T]] // SI-7818
   }
@@ -863,9 +863,9 @@ trait Observable[+T]
    * @see <a href="http://msdn.microsoft.com/en-us/library/system.reactive.linq.observable.window.aspx">MSDN: Observable.Window</a>
    */
   def sliding(timespan: Duration, timeshift: Duration, count: Int, scheduler: Scheduler): Observable[Observable[T]] = {
-    val span: Long = timespan.length
-    val shift: Long = timespan.unit.convert(timeshift.length, timeshift.unit)
-    val unit: TimeUnit = timespan.unit
+    val span: Long = timespan.toNanos
+    val shift: Long = timespan.toNanos
+    val unit: TimeUnit = duration.NANOSECONDS
     Observable.jObsOfJObsToScObsOfScObs(asJavaObservable.window(span, shift, unit, count, scheduler))
       : Observable[Observable[T]] // SI-7818
   }

